doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,278 @@
+# Reclocking implementation
+
+## Summary
+
+This document proposes a design for implementing reclocking and exposing the
+reclocked mappings to end users.
+
+## Goals
+
+* Describe the physical representation of the reclocking data.
+* Describe the policy for reclocking each type of source.
+* Describe how users interact with the reclocking data.
+
+## Non-Goals
+
+* Motivating reclocking. See the product spec in [#11218].
+* Describing reclocking formally. See the
+  [original reclocking design doc](20210714_reclocking.md).
+
+## Description
+
+### Reclockable source types
+
+We define a **reclockable** source type as one where the upstream system
+provides a useful [gauge of progress](20210714_reclocking.md#description).
+The following source types are reclockable:
+
+* File
+* Kafka
+* Kinesis
+* PostgreSQL
+* S3
+
+The following sources types are nonreclockable:
+
+* PubNub
```

Comment:
Why can't PubNub be reclocked?

Reply:
There is no upstream notion of progress, so no source gauge to reclock!